server/src/transport/ws.rs

```diff
@@ -426,31 +427,61 @@ where
 }
 
 /// Attempts to read data from WebSocket fails if the server was stopped.
-async fn try_recv<S>(receiver: &mut Receiver, data: &mut Vec<u8>, stopped: S) -> Receive<S>
+async fn try_recv<S>(receiver: &mut Receiver, data: &mut Vec<u8>, mut stopped: S, ping_config: PingConfig) -> Receive<S>
 where
 	S: Future<Output = ()> + Unpin,
 {
-	let receive = async {
-		// Identical loop to `soketto::receive_data` with debug logs for `Pong` frames.
-		loop {
-			match receiver.receive(data).await? {
-				soketto::Incoming::Data(d) => break Ok(d),
-				soketto::Incoming::Pong(_) => tracing::debug!("Received pong"),
-				soketto::Incoming::Closed(_) => {
-					// The closing reason is already logged by `soketto` trace log level.
-					// Return the `Closed` error to avoid logging unnecessary warnings on clean shutdown.
-					break Err(SokettoError::Closed);
-				}
-			}
+	let mut last_active = Instant::now();
+
+	let receive = futures_util::stream::unfold((receiver, data), |(receiver, data)| async {
+		match receiver.receive(data).await {
+			Ok(soketto::Incoming::Data(_)) => None,
+			Ok(soketto::Incoming::Pong(_)) => Some((Ok(()), (receiver, data))),
+			Ok(soketto::Incoming::Closed(_)) => Some((Err(SokettoError::Closed), (receiver, data))),
+			// The closing reason is already logged by `soketto` trace log level.
+			// Return the `Closed` error to avoid logging unnecessary warnings on clean shutdown.
+			Err(e) => Some((Err(e), (receiver, data))),
 		}
-	};
+	});
+
+	let mut inactivity_check =
+		Box::pin(ping_config.inactive_limit().map(|d| tokio::time::sleep(d).fuse()).unwrap_or_else(Fuse::terminated));
```

Comment:
I'm not super excited about `Box::pin(tokio::time::sleep(...))`, would be nicer with `tokio::time::Interval (which measures the time between ticks)` then we could just `reset` the interval if a message is received and once a `tick` expired then we know that `max_limit` has expired and just terminate without any extra checks.

However, that would require to set `tokio::time::interval(Duration::MAX)` when the inactivity check is disabled which I'm fine with

Reply:
Well, yes, that would have been more ergonomic way of doing things, but with Interval::tick, one encounters problems with borrow checker, as `interval` would be borrowed in `futures::select`, it would be impossible to call `reset` on it. I've rewritten it using `tokio::time::interval`, but still with extra checks because of borrow checker problems.